Chardev: refactor chardev creation, for better multiple backend support

vivianQizhu opened the pull request:
1. qdevices: Add Serial Bus representation
2. qcontainer: Add a wrapper for serial device creation

id: 1588423
Signed-off-by: qizhu <qizhu@redhat.com>

vivianQizhu commented:
Generate correct command line:
```
    -chardev socket,id=qmp_id_qmpmonitor1,path=/var/tmp/avocado_7boIC5/monitor-qmpmonitor1-20180703-134454-w7aI0QFb,server,nowait \
    -mon chardev=qmp_id_qmpmonitor1,mode=control  \
    -chardev socket,id=qmp_id_catch_monitor,path=/var/tmp/avocado_7boIC5/monitor-catch_monitor-20180703-134454-w7aI0QFb,server,nowait \
    -mon chardev=qmp_id_catch_monitor,mode=control \
    -device pvpanic,ioport=0x505,id=idVzkC1q \
    -chardev socket,id=chardev_serial0,host=127.0.0.1,port=5000,server,nowait \
    -device isa-serial,id=serial0,chardev=chardev_serial0 \
    -chardev socket,id=chardev_vc1,host=127.0.0.1,port=5001,server,nowait \
    -device virtio-serial-pci,id=virtio_serial_pci0,bus=pci.0,addr=0x3 \
    -device virtconsole,id=vc1,name=vc1,chardev=chardev_vc1,bus=virtio_serial_pci0.0,nr=1 \
    -chardev socket,id=chardev_vc2,host=127.0.0.1,port=5002,server,nowait \
    -device virtconsole,id=vc2,name=vc2,chardev=chardev_vc2,bus=virtio_serial_pci0.0,nr=2 \
    -chardev socket,id=chardev_vc3,host=127.0.0.1,port=5003,server,nowait \
    -device virtconsole,id=vc3,name=vc3,chardev=chardev_vc3,bus=virtio_serial_pci0.0,nr=3 \
    -chardev socket,id=chardev_vc4,host=127.0.0.1,port=5004,server,nowait \
    -device virtconsole,id=vc4,name=vc4,chardev=chardev_vc4,bus=virtio_serial_pci0.0,nr=4 \
    -chardev socket,id=chardev_vs1,host=127.0.0.1,port=5005,server,nowait \
    -device virtserialport,id=vs1,name=vs1,chardev=chardev_vs1,bus=virtio_serial_pci0.0,nr=5 \
    -chardev socket,id=chardev_vs2,host=127.0.0.1,port=5006,server,nowait \
    -device virtserialport,id=vs2,name=vs2,chardev=chardev_vs2,bus=virtio_serial_pci0.0,nr=6 \
    -chardev socket,id=chardev_vs3,host=127.0.0.1,port=5007,server,nowait \
    -device virtserialport,id=vs3,name=vs3,chardev=chardev_vs3,bus=virtio_serial_pci0.0,nr=7 \
    -chardev socket,id=chardev_vs4,host=127.0.0.1,port=5008,server,nowait \
    -device virtserialport,id=vs4,name=vs4,chardev=chardev_vs4,bus=virtio_serial_pci0.0,nr=8  \
```

xutian commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1666,6 +1666,163 @@ def images_define_by_params(self, name, image_params, media=None,
+            devices[-1].set_param('server', kwargs.get('server'))
+            devices[-1].set_param('nowait', kwargs.get('nowait'))
+            devices[-1].set_param('reconnect', kwargs.get('reconnect'))
+        devices[-1].format_params(devices[-1].params)
```
split lines from 1691 to 1729 into function  chardev_define_by_params.

vivianQizhu replied:
@xutian Here I can not go through kwargs and set all params inside it for `device[-1]`, since the kwargs actually contains some other options for other devices created later, not only for this device. So that's very hard to distinguish them earlier in `define_by_params`

vivianQizhu replied:
@xutian I have tried to move it to `serials_define_by_params`, only including chardev related options in **kwargs, and the options for other devices is laid outside as individual parameters. Thanks for your suggestions.

xutian commented on virttest/utils_misc.py:
```diff
@@ -440,14 +440,16 @@ def is_port_free(port, address):
 
 
-def find_free_port(start_port, end_port, address="localhost"):
+def find_free_port(start_port, end_port, address="localhost", reserved=[]):
```
don't used empty list as default value.

vivianQizhu replied:
@xutian Updated default value to `None`, thanks.

xutian commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1703,6 +1703,154 @@ def images_define_by_params(self, name, image_params, media=None,
+        serial_type = serial_params.get('serial_type', 'virtio')
+        port_type = None
+        machine_type = serial_params.get('machine_type')
+        if serial_type == 'virtio':  # translate virtio to ccw/device
```
line 1790 -  1806 are useless,  it should be instead with configuration in machine.cfg

sitoliu commented on virttest/qemu_devices/qdevices.py:
```diff
@@ -2064,11 +2065,28 @@ def _dev2addr(self, device):
         return [bus, unit]
 
 
+class QSerialBus(QBusUnitBus):
```
Hi @vivianzhu, would you mind explaining the difference between QBusUnitBus and QPCIBus? Why serial bus (do you mean `virtio-serial-pci`?) should be a QBusUnitBus not a QPCIBus? Thanks!

vivianQizhu commented:
@luckyh @yanan-fu I have updated it per your comments, please help review, thanks.

luckyh commented:
See also https://github.com/avocado-framework/avocado-vt/pull/1768

vivianQizhu commented:
> See also #1768

Yeah, the failure there is caused by the confounded 'id' and 'name',  the issue will be fixed by the wrapper defined here.

rayx commented:
So, is my PR a duplicate effort again? I wonder if someone in the maintainers can put the overall test status on a wiki, with information like which tests don't work and the related issue number or PR number. That would help people who are outside redhat to use the test and contribute to it more efficiently, I guess.

vivianQizhu commented:
@rayx there are serials of issue on virtio console test, we are working on it. FYI, https://github.com/autotest/tp-qemu/pull/1154 https://github.com/autotest/tp-qemu/pull/1037

rayx commented:
@vivianQizhu OK, thanks for information. I'll leave it to your folks and perhaps try other tests (if I still have the motivation :)

luckyh commented:
@rayx sorry for that. For tp-qemu issues, currently we are using redhat internal bugzilla tool do the management (because of some legacy reasons), and yes it is quite unfriendly to upstream users. Well, there are about hundreds of issues in the list (since vt tests support extremely large set of variants), it is not possible to mirror them all to github.

You just raised a good suggestion, I'll see if we can do some improvement for it. Thanks for your contributions!

balamuruhans commented:
@rayx you can submit an issue in this github repo when you start working on new feature implementation or fixing any issue observed. I helps us to discuss to find existing solution or PR sent, it can avoid this situation.
Your contribution will really helps us in making framework better.Thank you :)

rayx commented:
Thank @luckyh and @balamuruhans for your reply. I was probably a bit optimistic and underestimated the complexity and size of qemu test. My purpose is really about finding aarch64 specific failures. I'll focus on those issues and log the common issues I find on github.

luckyh commented:
@rayx ok, I see. @ldoktor often runs aarch64 specific tests, so I think he is the right person to tell about the status of that, thanks.

vivianQizhu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1720,6 +1720,147 @@ def images_define_by_params(self, name, image_params, media=None,
+        :return: the device list that construct the serial device
+        """
+        if (not self.has_option('chardev') or
+                not self.has_device(serial_type)):
```
@yanan-fu I think it should get `True` for `not self.has_device(serial_type)`, for the default definition `serial_type = "isa-serial"`, so it get `-serial` by L1743

yanan-fu replied:
@vivianQizhu  `I think it should get True for not self.has_device(serial_type)`  ?   do you mean are do not have `isa-serial` device ?

The old logic is,  arm have `isa_serial` but it can not work normally, so use `- serial` instead.

vivianQizhu replied:
@yanan-fu Really? Thanks for the information, so is that a workaround for the product issue? I will modify the code to handle it if so.

yanan-fu replied:
@vivianQizhu  please refer the old comment, and if we do not test with other arch,  we should keep old logic.

vivianQizhu replied:
@yanan-fu You are right, I will update it. Thanks.

vivianQizhu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1720,6 +1720,147 @@ def images_define_by_params(self, name, image_params, media=None,
+        if (not self.has_option('chardev') or
+                not self.has_device(serial_type)):
+            cmd = " -serial unix:'%s',server,nowait" % file_name
+            return qdevices.StrDev('SER-%s' % serial_id, cmdline=cmd)
```
@yanan-fu Thanks, it should be `qdevices.QStringDevice`, `StrDev = qdevices.QStringDevice` is defined in `qemu_vm` but we are in `qcontainer`, will update.

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1720,6 +1720,147 @@ def images_define_by_params(self, name, image_params, media=None,
+                        'logappend': serial_params.get('chardev_logappend')}
+        backend = serial_params.get('chardev_backend', 'unix_socket')
+        if backend not in ['socket', 'vc', 'ringbuf',
+                           'spiceport', 'spicevmc']:
```
`socket` and  `vc`  have option `mux`, why exclude them ?

yanan-fu replied:
Default value for `backend` is `unix_socket`,  but here it use `['socket', 'vc', 'ringbuf','spiceport', 'spicevmc']`.

vivianQizhu replied:
@yanan-fu That's the condition statement for the option `mux`, nothing to do with what is default value.

vivianQizhu replied:
@yanan-fu Ah I get your point, `socket` should be `unix_socket`, but yes,  `socket` and `vc` have the option, should be removed from this list. I just copy it from CharDevice, not confirm all of them, my careless. Thank you for being so careful.

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1720,6 +1720,147 @@ def images_define_by_params(self, name, image_params, media=None,
+        elif backend in ['stdio']:
+            chardev_args.update({'signal': serial_params.get('chardev_signal')})
+        elif 'unix' in backend:  # unix_socket
+            chardev_args.update({'path': file_name})
```
Can move it together with L1827~1829 ( when use tcp_socket, unix_socket)

yanan-fu commented on virttest/qemu_devices/qdevices.py:
```diff
@@ -2091,11 +2092,28 @@ def _dev2addr(self, device):
+        """
+
+        super(QSerialBus, self).__init__(busid, 'SERIAL', [1, max_ports],
+                                         aobject, bus_type, unit_spec='nr')
```
Could you explain which device has option `nr` ?   I can't get it from qemu side.

vivianQizhu replied:
@yanan-fu 
```
#  /usr/libexec/qemu-kvm -device virtserialport,?
virtserialport.name=str
virtserialport.chardev=str (ID of a chardev to use as a backend)
virtserialport.nr=uint32
```
That should the virtio-serial bus distributing to each device.

yanan-fu commented on virttest/qemu_devices/qdevices.py:
```diff
@@ -2039,7 +2037,7 @@ class QBusUnitBus(QDenseBus):
 
     """ Implementation of bus-unit bus (ahci, ide) """
```
Need update this doc, since it is not only for 'unit' option now,  then not only ahci and ide.

vivianQizhu replied:
@yanan-fu I would like to keep the description in QSerialBus instead of here, the class QBusUnitBus can not represent SerialBus directly, the inherited relationship is sufficient to explain their relation.

yanan-fu replied:
@vivianQizhu  My meaning is `bus-unit` is not suitable, as it can be `nr` too.
And for `ahci, ide`,   there also have  `QAHCIBus` and `QIDEBus`.  I think there is not conflicts with `QBusUnitBus`  and `QSerialBus`.   We can add docs for both of them. What do you think ?

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1720,6 +1720,147 @@ def images_define_by_params(self, name, image_params, media=None,
+        if 'socket' in backend:  # tcp_socket & unix_socket
+            chardev_args.update({'server': serial_params.get('chardev_server', 'yes'),
+                                 'nowait': serial_params.get('chardev_nowait', 'yes'),
+                                 'reconnect': serial_params.get('chardev_reconnect')})
```
New design: param name is `chardev_$option_name`.
Why don't use class CharDevice directly to generate chardev_args ?  (improve it to match the new request)
In this way,  param name should be same with option name.
```
1012         options = self.get_supported_options(backend)                           
1013         params = params.copy_from_keys(options)  
```

vivianQizhu replied:
@yanan-fu The method you mentioned is not recommended by other maintainer, and it is indeed not flexible enough for all conditions. I have no strong opinion on this, just follow the way how we create image.

yanan-fu replied:
@vivianQizhu  In my option,  L1821~1852 are almost duplicate with qdevices.CharDevice. 
For image, there do not have similar function to do this,   but for chardev, it have.

I also think about delete this part (get_supported_options, copy_from_keys, format_params) from CharDevice, but we can since it is useful for hotplug/ hot-unplug. Some other cases use it.

If use the current design, i am afraid we have to change both of these two parts if qemu have change in the future.

vivianQizhu replied:
@yanan-fu All the chardev device should be created from the new introduced wrappers in the further, so there will be no other case use it. chardev usually is a set of devices, it can not be created alone. That's why it is similar with image: a complete device is composed by a set of device(bus/backend/frontend). We need to handle other factors related to environment besides options, like machine types, and the relation between devices, which are not proper to handle in device object, which means, CharDevice is not able to handle all the options. Thus, we have to define a wrapper anyway. 
In conclusion, the options of the there part of a complete device should be handled together, so not in device level.

vivianQizhu replied:
@yanan-fu And, for other device that use chardev as backend(like nic), they should handle the options in its device creation part, not inside CharDevice object either.  So here we take a real device as an entirety in our design, not a single command line.

yanan-fu replied:
@vivianQizhu  L1835~L1863 have nothing to do with machine type and others, it is only related with chardev options,  and ChaDevice.get_supported_options do the same thing( together with  copy_from_keys).  Correct me if i am wrong.

For this question:  `so there will be no other case use it`:  CharDevice support hotplug/hot-unplug,  `qemu_guest_agent_hotplug` use it.  How can we run hotplug without CharDevice ?  

If no CharDevice, current implement is good for me,  but why don't we call it directly while we can  not delete it.

vivianQizhu replied:
@yanan-fu 
My point is, they should be handled together with all related devices, not in QDevice object level.
For hotplug, it should be all replaced with simple_hotplug().

vivianQizhu replied:
@yanan-fu Well, actually I was about to only use CharDevice at the beginning. @luckyh Would you help to share your thought here? Thanks.

vivianQizhu commented on virttest/qemu_devices/qdevices.py:
```diff
@@ -2054,8 +2052,9 @@ def __init__(self, busid, bus_type, lengths, aobject=None, atype=None):
         """
```
@yanan-fu Not sure what's wrong here, but I am not able to modify it, it looks normal in my local environment...

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1720,6 +1720,150 @@ def images_define_by_params(self, name, image_params, media=None,
+        if 'socket' in backend:  # tcp_socket & unix_socket
+            chardev_args.update({'server': serial_params.get('chardev_server', 'yes'),
+                                 'nowait': serial_params.get('chardev_nowait', 'yes'),
+                                 'reconnect': serial_params.get('chardev_reconnect')})
```
@vivianQizhu  Checked all the options for `-chardev`.
1. For backend `ringbuf`, miss `size`.
2. For backend `unix_socket`, miss `telnet`.   and miss `websocket` only for qemu-3.1.0.
3. For backend `tcp_socket`, miss `telnet`, `tls-creds`, and miss  `websocket` only for qemu-3.1.0.
4. For backend `vc`, miss `[[,width=width][,height=height]][[,cols=cols][,rows=rows]]`.

I do not know if there are cases test these options, ignore me if no need to add them specially.

vivianQizhu replied:
@yanan-fu `telnet` and `tls-creds` should be included, for others there are no requirements currently. I will update it for `telnet` and `tls-creds` at least. Thanks.

vivianQizhu replied:
@yanan-fu I am not adding `tls-creds` for now, the test case related to it requires complex configuration and in low priority, we have no plan to automate it recently.

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1720,6 +1720,150 @@ def images_define_by_params(self, name, image_params, media=None,
+        elif backend == 'stdio':
+            chardev_args.update({'signal': serial_params.get('chardev_signal')})
+        elif backend in ['udp', 'tcp_socket']:
+            chardev_args.update({'host': serial_params.get('chardev_host'),
```
these  3 `elif` can be `if` ?

vivianQizhu replied:
@yanan-fu `elif` is more efficient than `if`

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1720,6 +1720,150 @@ def images_define_by_params(self, name, image_params, media=None,
+            else:
+                controller_suffix = 'pci'
+            serial_type = '%s-%s' % (serial_type, controller_suffix)
+            virtio_port_type = "virt%s" % virtio_port_type
```
how about use `virtserialport`, `virtconsole` as `virtio_port_type` ?  It is more easy to understand as it is same with qemu help output.  And then no need to add prefix `virt` here.

vivianQizhu replied:
@yanan-fu `virtio_port_type` is previously defined as `serialport` and `console`, just want to adopt previous usage.

yanan-fu replied:
@vivianQizhu  I understand your point. 
Previously,  `console`, `virtio_console`, `serialport`, `virtio_serial` all can use,  since framwork accept all of them, and convert them to acceptable port type of qemu command line.
```
port_params.get('virtio_port_type') in ("console", "virtio_console"):
```
```
if 'console' in params.get('virtio_port_type'):
                port_type = 'virtconsole'
            else:
                port_type = 'virtserialport'
```

For me, it is confused.   If `console` , `virtio_console` can only be `virtconsole`,  and `serialport`, `virtio_serial` can only be `virtserialport`,  use the final port type is more clearly.
While in your another patch https://github.com/autotest/tp-qemu/pull/1434,  you have correct `virtio_serial`  to `serialport`, i think we can update them together.

vivianQizhu replied:
@yanan-fu Yes, it makes sense, I will update them.

vivianQizhu replied:
@yanan-fu There is variants named `virtserialport` and `virtconsole` in test case, I am afraid there is name conflict if modified them, I will need rename the variants and reorganize those `only` `no` keyword for them. I think leave them there is better, not really necessary to reorganize all params, which makes no improvement, just for save this postfix adding. Thanks.

yanan-fu replied:
@vivianQizhu  oh~  i see those variants,  if so, it is ok to use `console` and `serialport`.

yanan-fu commented on virttest/qemu_devices/qdevices.py:
```diff
@@ -1009,8 +1009,6 @@ def __init__(self,
         self.verify_supported_backend(backend)
-        options = self.get_supported_options(backend)
-        params = params.copy_from_keys(options)
         params = self.format_params(params)
```
After remove `params = params.copy_from_keys(options)`,  `params=None` here, so it is meaningless for `format_params`.

vivianQizhu replied:
@yanan-fu Actually they are independent, `format_params()` translate params before `yes` `no` `Ture` `False`, which is still in required here.

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1720,6 +1720,150 @@ def images_define_by_params(self, name, image_params, media=None,
+            return legacy_dev
+
+        virtio_port_type = serial_params.get('virtio_port_type')
+        if virtio_port_type:
```
In qemu-vm, we already use:
```
            virtio_port_type = serial_params.get('virtio_port_type')
            if virtio_port_type:
                serial_params['serial_type'] = 'virtio-serial'
```
and have got `serial_type` from above, so how about  `if serial_type.startswith('virtio')`  or `if serial_type == 'virtio-serial'` ?   and i prefer the second one.

vivianQizhu replied:
@yanan-fu And for here, qemu-vm is not the only user for the API, we must make sure `virtio_port_type` is defined for `virtio-serial` type anytime when user call it, e.g. define a new device for hotplug.

yanan-fu replied:
@vivianQizhu  I have different option,  the `serial_type` is depends on `virtio_port_type` all the time with the current implement, in this API,  it already use `L1798    serial_type = serial_params['serial_type']`.

I am not very clear about other usage now, but if it is a problem,  i think we need to update `serial_type` first.

yanan-fu commented:
Need replace `virtio_port_chardev` with `chardev_backend` in base.cfg, although it is a comment.

vivianQizhu commented:
@yanan-fu Updated as we discussed offline, please help review, thanks.

vivianQizhu commented:
Updated with:
1. Remove param `virtio_port_type`, use single `serial_type` to represent virtconsole and virtserialport.
2. Removed most unless chardev options, keep the original filter in the object `CharDevice`, so pass `Params()` instead of general dictionary when initialize it.
3. Initialize a new bus for virtconsole since they can not share a single bus, which I missed previously.
Also update the tp-qemu PR together and generally test it generates correct command line.
@luckyh @yanan-fu Please help review again, thanks.

yanan-fu commented on virttest/qemu_vm.py:
```diff
@@ -2786,7 +2633,9 @@ def create(self, name=None, params=None, root_dir=None,
             for serial in params.objects("serials"):
-                self.serial_ports.append(serial)
+                if params.object_params(serial).get(
+                        "serial_type") != 'virtio-serial':
```
We don't use `virtio-serial` now,  we use `virtserialport`, `virtconsole` and so on.

yanan-fu commented on virttest/qemu_vm.py:
```diff
@@ -2390,41 +2259,19 @@ def _create_serial_console(self):
                                                log_name)
         self.serial_console = aexpect.ShellSession(
-            "nc -U %s" % self.get_serial_console_filename(tmp_serial),
+            "nc -U %s" % self.serial_console_log,
```
`serial_console_log` is the log path + log file name, it is used to save the serial log.
`get_serial_console_filename(serial_session_device)` is the `path` in qemu command line.
What we can use for 'nc' is the `path` in qemu command line,  not log file name.

vivianQizhu commented on virttest/qemu_vm.py:
```diff
@@ -1613,83 +1516,49 @@ def sort_key(dev):
+        serials = params.objects('serials')
+        if serials:
+            self.serial_session_device = serials[0]
+            host = params.get('serial_chardev_host', '127.0.0.1')
```
@yanan-fu Honestly I don't want support that, that's too complicated and meaningless from testing perspective, it is only about socket transmission, same host or multi-host has no significant difference, so I will update the example in base.cfg. Thanks.

yanan-fu replied:
`chardev_host`

vivianQizhu commented:
@yanan-fu Thanks for review, updated.

vivianQizhu commented:
@yanan-fu I have updated it with a new function `chardev_define_by_params`, please help review, thanks.

vivianQizhu commented on shared/cfg/base.cfg:
```diff
@@ -129,22 +129,22 @@ q35, arm64-pci, arm64-mmio:
+# serial_chardev_backend_vs1 = unix_socket
+# serial_chardev_backend_vc1 = tcp_socket
+# Chardev host for tcp/udp socket backend
+# serial_chardev_host = $host_ip (127.0.01 by default)
```
@yanan-fu Updated, thanks.

vivianQizhu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1718,6 +1719,146 @@ def images_define_by_params(self, name, image_params, media=None,
+                 'nowait': params.get('chardev_nowait', 'yes')})
+        elif backend in ['spicevmc', 'spiceport']:
+            chardev_param.update({'debug': params.get('chardev_debug'),
+                                 'name':  params.get('serial_name')})
```
@yanan-fu Should be `chardev_name`, will update, thanks.

yanan-fu replied:
And no call for `chardev_name`  from `params['chardev_name'] = name`,  do you mean  `chardev_name` ?

vivianQizhu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1718,6 +1719,146 @@ def images_define_by_params(self, name, image_params, media=None,
+
+        prefix = params.get('virtio_port_name_prefix')
+        name = serial_id
+        name = prefix + name if prefix else name
```
@yanan-fu Updated, thanks.

yanan-fu commented on shared/cfg/base.cfg:
```diff
@@ -129,22 +129,22 @@ q35, arm64-pci, arm64-mmio:
 # virtio_port_name_prefix_vc1 = ""
 
+serials = "serial0"
+serial_type = "isa-serial"
```
To avoid  other operation in case level affect the default serial session by mistake,  how do you think about use fixed `serial_type_serial0 = isa-serial` ?  
And in qemu_vm,  use `self.serial_session_device = serial0` instead of `serials[0]`
```
        if serials:
            self.serial_session_device = serials[0]
```

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1718,6 +1719,145 @@ def images_define_by_params(self, name, image_params, media=None,
+        :param file_name: file name of chardev (optional)
+        :return: CharDevice object
+        """
+        backend = params.get('chardev_backend', 'unix_socket')
```
No param `chardev_backend` now.
For serial, we have `serial_chardev_backend`,  if want to distinguish other device,  we may need make `backend` as a param of  `chardev_define_by_params` .

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1718,6 +1719,145 @@ def images_define_by_params(self, name, image_params, media=None,
+                {'host': params['chardev_host'],
+                 'port': params['chardev_port'],
+                 'ipv4': params.get('chardev_ipv4'),
+                 'ipv6': params.get('chardev_ipv6')})
```
This 4 params have the same problem with `chardev_backend`.  Especially when use `params[***]`, no `get(***)`

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1718,6 +1719,145 @@ def images_define_by_params(self, name, image_params, media=None,
+                 'nowait': params.get('chardev_nowait', 'yes')})
+        elif backend in ['spicevmc', 'spiceport']:
+            chardev_param.update({'debug': params.get('chardev_debug'),
+                                 'name':  params.get('chardev_name')})
```
Failed as:   'qemu-kvm: -chardev spicevmc,id=chardev_vs3,name=com.redhat.spice.vs3: unsupported type name: com.redhat.spice.vs3
allowed spice char type names: vdagent, smartcard, usbredir

vivianQizhu commented:
@yanan-fu Updated,
1. Remove some `serial_` prefix.
2. Line https://github.com/avocado-framework/avocado-vt/pull/1639/files#diff-4045499c9c61cb3d07734c53a7dcbb26R1814 , corresponding to https://github.com/avocado-framework/avocado-vt/pull/1639/files#diff-3f2317089f10636f453dca6c4f880f6cL1701, to resolve the spice issue. It use same value of serial id(`port_name` previously) for `name`, let's just keep the original design, but also allowing customer it by defining `chardev_name`.

yanan-fu commented on virttest/qemu_vm.py:
```diff
@@ -1632,83 +1535,47 @@ def sort_key(dev):
+                reg = 0x30000000 + 0x1000 * reg_count
+                serial_params['serial_chardev_reg'] = hex(reg)
+                reg_count += 1
+            backend = serial_params.get('serial_chardev_backend',
```
`chardev_backend`

yanan-fu commented on virttest/qemu_vm.py:
```diff
@@ -1632,83 +1535,47 @@ def sort_key(dev):
+            if 'ppc' in params.get('vm_arch_name', arch.ARCH)\
+                    and serial_params.get('serial_type') == 'spapr-vty':
+                reg = 0x30000000 + 0x1000 * reg_count
+                serial_params['serial_chardev_reg'] = hex(reg)
```
`chardev_reg`

vivianQizhu replied:
@yanan-fu Updated, but this one is `serial_reg`, it's an option for serial device, not chardev.

yanan-fu commented on virttest/qemu_vm.py:
```diff
@@ -1632,83 +1535,47 @@ def sort_key(dev):
+                                        'unix_socket')
+            if backend in ['udp', 'tcp_socket']:
+                serial_params['serial_chardev_host'] = host
+                serial_params['serial_chardev_port'] = free_ports[index]
```
`chardev_host` and `chardev_port`

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1718,6 +1719,146 @@ def images_define_by_params(self, name, image_params, media=None,
+        elif backend in ['spicevmc', 'spiceport']:
+            chardev_param.update(
+                {'debug': params.get('chardev_debug'),
+                 'name': params.get('chardev_name')})
```
same with before, test failed.

qemu command line:
```
    -chardev spicevmc,id=chardev_vs3,name=vs3 \
    -device virtserialport,id=vs3,name=com.redhat.spice.vs3,chardev=chardev_vs3,bus=virtio_serial_pci0.0,nr=3 \
```

failed reason:
```
qemu-kvm: -chardev spicevmc,id=chardev_vs3,name=vs3: unsupported type name: vs3
allowed spice char type names: vdagent, smartcard, usbredir'
```
I am not familiar with this,  `chardev_name` is must for `spicevmc` ?

vivianQizhu replied:
@yanan-fu Please refer here: https://github.com/autotest/tp-qemu/pull/1434/files#diff-8d0a580773b347958522648c9657235eL10 should be `vdagent`

yanan-fu replied:
@vivianQizhu  Got the point.  The value of `name` option for `chardev` can only be `vdagent, smartcard, usbredir`.

In this way,  i think `params['chardev_name'] = params.get('chardev_name', serial_id)` is unreasonable.
If user want to test backend `spicevmc`,  `chardev_name` should be necessary, user should give correct value,  othervise, case failed as `qemu-kvm: -chardev spicevmc,id=chardev_vs3: chardev: spice channel: no name given'`

vivianQizhu replied:
@yanan-fu I think we better adopt the original usage(no `chardev_name`): https://github.com/avocado-framework/avocado-vt/pull/1639/files#diff-3f2317089f10636f453dca6c4f880f6cL1701, so by default user only need to define `serial_id` as `vdagent, smartcard, usbredir`, like previously.

vivianQizhu replied:
@yanan-fu Removed `params['chardev_name'] = params.get('chardev_name', serial_id)`

yanan-fu commented on virttest/qemu_devices/qcontainer.py:
```diff
@@ -1718,6 +1719,145 @@ def images_define_by_params(self, name, image_params, media=None,
 
+    def serials_define_by_variables(self, serial_id, serial_type, chardev_id,
+                                    bus_type=None, serial_name=None,
+                                    bus=None, nr=None, reg=None):
```
Test pass now.
 Only one concern, why do not keep consistent with other serial option, i mean, `serial_name`, but not `serial_reg`, `serial_nr` ?

yanan-fu commented on shared/cfg/base.cfg:
```diff
@@ -129,22 +129,25 @@ q35, arm64-pci, arm64-mmio:
+# Chardev host for tcp/udp socket backend
+# chardev_host = $host_ip (127.0.0.1 by default)
+# chardev_ipv4 = yes
+# chardev_ipv6 = yes
```
By default,  no `ipv4` and `ipv6`. It is
```
                 'ipv4': params.get('chardev_ipv4'),
                 'ipv6': params.get('chardev_ipv6')})
```
qemu command line `-chardev udp,host=127.0.0.1,id=chardev_vc2,port=5746`

vivianQizhu replied:
@yanan-fu It's just an example to show how to use it, not mean default here.

yanan-fu replied:
@vivianQizhu  ok, for usage, `yes` `on` `True` all can use:
```
        for opt in ["server", "telnet", "nowait",
                    "ipv4", "ipv6", "nodelay"]:
            if params.get(opt) in ["yes", "on", True]:
                params[opt] = "NO_EQUAL_STRING"
            elif opt in params:
                del params[opt]
```

vivianQizhu replied:
@yanan-fu I want to keep it simple, so only list one usage here, to not disturb users.

yanan-fu replied:
It is ok for me ~

yanan-fu commented on virttest/qemu_vm.py:
```diff
@@ -1632,83 +1535,47 @@ def sort_key(dev):
+            reg_count = 0
+        self.virtio_ports = []
+        for index, serial in enumerate(serials):
             serial_filename = vm.get_serial_console_filename(serial)
```
Since  isa-serial  and  virtio serial use same method to get `serial_filename` now. I wonder if we need to update  function `get_virtio_port_filename` in `virt_vm.py`.  Since it use  `virtio_port-%s-%s" % (port_name, self.instance)`. 

It may do not affect our test, as in qemu_vm  `get_virtio_port_filename`  is ok.

yanan-fu replied:
@vivianQizhu   correct my comment above,   qemu_vm use `get_virtio_port_filenames`  not `get_virtio_port_filename`.   only one `get_virtio_port_filename` in virt_vm.  It is not correct now.

vivianQizhu replied:
@yanan-fu I see the two impact ones: https://github.com/autotest/tp-qemu/blob/master/qemu/tests/qemu_guest_agent.py#L199 and https://github.com/autotest/tp-qemu/blob/master/qemu/tests/qemu_guest_agent_hotplug.py#L32
What do you think if I replace them with `get_serial_console_filename`? Cause now we actually totally abandon the usage of `get_virtio_port_filename` in qemu_vm context.

vivianQizhu replied:
@yanan-fu And for qemu_vm.py:
```
diff --git a/virttest/qemu_vm.py b/virttest/qemu_vm.py
index 9fc849c9..b0c51f62 100644
--- a/virttest/qemu_vm.py
+++ b/virttest/qemu_vm.py
@@ -3018,7 +3018,6 @@ class VM(virt_vm.BaseVM):
         # Generate the tmp file which should be deleted.
         file_list = [self.get_testlog_filename()]
         file_list += qemu_monitor.get_monitor_filenames(self)
-        file_list += self.get_virtio_port_filenames()
         file_list += self.get_serial_console_filenames()
         file_list += list(self.logs.values())
```

yanan-fu replied:
@vivianQizhu  it is ok for me to use `get_serial_console_filename` in case level. as when generate qemu command line, we use same method to generate filename.   
And  remove `get_virtio_port_filename` and `get_virtio_file_names` directly ?   i have checked tp-libvirt, no use.

vivianQizhu replied:
@yanan-fu Even it is not used by tp-libvirt now, I don't really want to change the base class, I only care how qemu_vm works.

yanan-fu replied:
At least,  the one in qemu_vm need be deleted, `get_virtio_port_filenames`. 

For virt_vm, i prefer to delete the two function.  As my understanding, only qemu care it but do not need it.  

@xutian  do you have any idea about this issue ?

vivianQizhu replied:
@yanan-fu Okay, all deleted, hope has no side effect.

yanan-fu commented on virttest/qemu_vm.py:
```diff
@@ -142,6 +142,7 @@ def __init__(self, name, params, root_dir, address_cache, state=None):
             self.serial_ports = []
             self.serial_console_log = None
+            self.serial_session_device = None
             self.virtio_console = None
```
We still have `self.virtio_ports` now, but  function `create_virtio_console` be totally removed.

vivianQizhu replied:
@yanan-fu You mean `self.virtio_console`?

yanan-fu replied:
@vivianQizhu  we have `self.virtio_ports`,  but no `self.virtio_console` as function `create_virtio_console`  be removed.   `_create_serial_console` is only for `serials[0]`

yanan-fu replied:
As we discussed,  user can only connect to one console (the first serial) at the same time, use common `_create_serial_console`. So remove ` self.virtio_console = None` here is ok.

vivianQizhu commented:
@yanan-fu Update, removed `self.virtio_console`, please help review again, thanks.

yanan-fu commented:
Code LGTM, and  test pass,  can get expect qemu command line.

Test with 
```
    serials += " vs1 vs2 vs3 vc1 vc2 vc3"
    serial_type_vs1 = virtserialport
    serial_type_vs2 = virtserialport
    serial_type_vs3 = virtserialport
    serial_type_vc1 = virtconsole
    serial_type_vc2 = virtconsole
    serial_type_vc3 = virtconsole
    chardev_backend_vs1 = unix_socket
    chardev_backend_vs2 = udp
    chardev_backend_vs3 = spicevmc
    chardev_backend_vc1 = unix_socket
    chardev_backend_vc2 = udp
    chardev_backend_vc3 = spicevmc
    chardev_name_vc3 = vdagent
    chardev_name_vs3 = vdagent
    virtio_port_name_prefix_vc3 = "com.redhat.spice."
    virtio_port_name_prefix_vs3 = "com.redhat.spice."
```

qemu command line:
```
    -chardev socket,server,nowait,path=/var/tmp/avocado_o8q7_l_f/serial-serial0-20190626-004342-PWQHBbub,id=chardev_serial0 \
    -device isa-serial,id=serial0,chardev=chardev_serial0 \

    -chardev socket,server,nowait,path=/var/tmp/avocado_o8q7_l_f/serial-vs1-20190626-004342-PWQHBbub,id=chardev_vs1 \
    -device virtio-serial-pci,id=virtio_serial_pci0,bus=pci.0,addr=0x3 \
    -device virtserialport,id=vs1,name=vs1,chardev=chardev_vs1,bus=virtio_serial_pci0.0,nr=1 \

    -chardev udp,host=127.0.0.1,port=5599,id=chardev_vs2 \
    -device virtserialport,id=vs2,name=vs2,chardev=chardev_vs2,bus=virtio_serial_pci0.0,nr=2 \

    -chardev spicevmc,name=vdagent,id=chardev_vs3 \
    -device virtserialport,id=vs3,name=com.redhat.spice.vs3,chardev=chardev_vs3,bus=virtio_serial_pci0.0,nr=3 \

    -chardev socket,server,nowait,path=/var/tmp/avocado_o8q7_l_f/serial-vc1-20190626-004342-PWQHBbub,id=chardev_vc1 \
    -device virtio-serial-pci,id=virtio_serial_pci1,bus=pci.0,addr=0x4 \
    -device virtconsole,id=vc1,name=vc1,chardev=chardev_vc1,bus=virtio_serial_pci1.0,nr=1 \

    -chardev udp,host=127.0.0.1,port=5398,id=chardev_vc2 \
    -device virtio-serial-pci,id=virtio_serial_pci2,bus=pci.0,addr=0x5 \
    -device virtconsole,id=vc2,name=vc2,chardev=chardev_vc2,bus=virtio_serial_pci2.0,nr=1 \

    -chardev spicevmc,name=vdagent,id=chardev_vc3 \
    -device virtio-serial-pci,id=virtio_serial_pci3,bus=pci.0,addr=0x6 \
    -device virtconsole,id=vc3,name=com.redhat.spice.vc3,chardev=chardev_vc3,bus=virtio_serial_pci3.0,nr=1  \
```


Acked-by: Yanan Fu <yfu@redhat.com>

xutian commented on virttest/qemu_vm.py:
```diff
@@ -2442,43 +2308,22 @@ def _create_serial_console(self):
+            self.serial_session_device, self.name)
         self.serial_console_log = os.path.join(utils_misc.get_log_file_dir(),
                                                log_name)
+        file = self.get_serial_console_filename(self.serial_session_device)
```
bad idea to use key words as var name.

vivianQizhu replied:
@xutian Yes, updated, thanks.

xutian commented on virttest/qemu_vm.py:
```diff
@@ -1632,83 +1534,47 @@ def sort_key(dev):
+
+            # Create virtio_ports (virtio_serialports and virtio_consoles)
+            serial_type = serial_params['serial_type']
+            if serial_type.startswith('virt'):
```
For making code more readable, please use "virtio" instead "virt"

yanan-fu replied:
@xutian  serial_type is `virtserialport` and `virtconsole`, so we need use `virt` here :)

vivianQizhu replied:
@xutian They are `virtserialport` and `virtconsole`, do no start with `virtio`.

xutian replied:
OK, please fix in your comments in your code

xutian commented on virttest/qemu_vm.py:
```diff
@@ -1632,83 +1534,47 @@ def sort_key(dev):
+            free_ports = utils_misc.find_free_ports(
+                5000, 6000, len(serials), host)
+            reg_count = 0
+        self.virtio_ports = []
```
let's move this define in  function  '__init__'

vivianQizhu replied:
@xutian It actually has one in `__init__`, I removed this redundant one, thanks.

yanan-fu replied:
@xutian  oh~   `__inint__`  already have, we can remove this one.

vivianQizhu commented:
@xutian Updated the comments, thanks.

yanan-fu commented:
gating test pass.

Ack

vivianQizhu commented:
Thanks all.
